README.md
metrics
========

Scripts for gathering infrastructure metrics.

Nordstrom Fellows
Please see https://confluence.nordstrom.net/display/UCG/Metrics for more info.

Everyone Else
Please see https://kodywilson.com/

## Installation
Clone the repo, create reasonable json files, install required gems.

```sh
git clone https://git.nordstrom.net/scm/ucg/metrics.git
cd metrics
vim pure_stats.json
gem install stoarray influxdb
```

## Usage
space_tracer_xxxxxx.rb - captures capacity metrics
perf_tracer_xxxxxx.rb  - captures performance metrics

Test the scripts, put in cron. Comment out the csv and db writes for testing.
We run the perf scripts every minute and capacity every 30 minutes.

```sh
sudo crontab -e
```

### Example cron entries

```sh
# Gather stats from storage arrays
*/30 * * * * /opt/chef/embedded/bin/ruby /u01/app/prd/pure_stats/space_tracer_pure.rb > /dev/null 2>&1
*/30 * * * * /opt/chef/embedded/bin/ruby /u01/app/prd/vmax_stats/space_tracer_vmax.rb > /dev/null 2>&1
*/1 * * * * /opt/chef/embedded/bin/ruby /u01/app/prd/pure_stats/perf_tracer_pure.rb > /dev/null 2>&1
```

## JSON configuration file examples
---------------------------

### Pure:
```json
    {
      "base_dir": "/u01/app/prd/pure_stats/",
      "database": "nameofyourdatabase",
      "db_host": "databasehost.nordstrom.net",
      "db_user": "putdatabaseuserhere",
      "db_pass": "base64encodeddatabaseuserspassword",
      "headers": { "Content-Type": "application/json" },
      "arrays": {
        "array01": "BASE64ENDCODEDPUREAPITOKEN4ARRAY01XXXXXXXXXXXXXX",
        "array02": "BASE64ENDCODEDPUREAPITOKEN4ARRAY02XXXXXXXXXXXXXX",
        "array03": "BASE64ENDCODEDPUREAPITOKEN4ARRAY03XXXXXXXXXXXXXX",
        "array04": "BASE64ENDCODEDPUREAPITOKEN4ARRAY04XXXXXXXXXXXXXX"
      }
    }
```

+ base_dir - Directory to start in.
+ database - Name of the database you are connecting.
+ db_host  - Fully qualified host name of database server.
+ db_user  - Database user.
+ db_pass  - Database user's password.
+ arrays   - Name of array with base64 encoded token.

### Vmax:
```json
    {
      "base_dir": "/u01/app/prd/vmax_stats/",
      "database": "nameofyourdatabase",
      "db_host": "databasehost.nordstrom.net",
      "db_user": "putdatabaseuserhere",
      "db_pass": "base64encodeddatabaseuserspassword",
      "headers": { "Content-Type": "application/json" }
    }
```

### Xtremio:
```json
    {
      "base_dir": "/u01/app/prd/xtremio_stats/",
      "database": "nameofyourdatabase",
      "db_host": "databasehost.nordstrom.net",
      "db_user": "putdatabaseuserhere",
      "db_pass": "base64encodeddatabaseuserspassword",
      "headers": {
        "Content-Type": "application/json",
        "Accept": "application/json"
      },
      "arrays": {
        "xtremio01": "Basic base64encodeuser:passforxtremio01",
        "xtremio02": "Basic base64encodeuser:passforxtremio02",
        "xtremio03": "Basic base64encodeuser:passforxtremio03"
      }
    }
```

## Troubleshooting

Nothing to mention yet.

## Development

Clone, hack, submit a pull request!

## Contributing

Bug reports and pull requests are welcome on Stash at https://git.nordstrom.net/scm/ucg/metrics.git. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The scripts are available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# CHANGELOG for metrics

This file is used to list changes made to the metrics repo.

## 11032016

* Removed influxxer scripts as they are no longer needed.
* Added link to further information on using these scripts.
* Commented csv creation line in space_tracer_pure.rb

## 08102016

* General cleanup, lowered sleeps, finally settled on conf file handling.
* Added link to Confluence documentation in README.

## 08032016

* General cleanup, lowered sleeps, finally settled on conf file handling.
* Added link to Confluence documentation in README.
* Created and tested perf_tracer_vmax.rb.

## 07252016

* Created and tested perf_tracer_xtremio.rb.
* Updated README with general script descriptions.

## 07222016

* Created and tested space_tracer_xtremio.rb.
* Updated README with Xtremio example json.

## 07202016

* Initial commit.

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


storage/influxxer_pure.rb
#!/usr/bin/env ruby
# Processes csv files containing historical Pure array capacity data.

require 'csv'
require 'json'
require 'influxdb'

conf     = JSON.parse(File.read('/u01/app/prd/pure_stats/pure_stats.json'))
base_dir = conf['base_dir']
database = conf['database']
host     = conf['db_host']
username = conf['db_user']
password = Base64.decode64(conf['db_pass'])
measure  = 'capacity' # The metric

influxdb = InfluxDB::Client.new database, username: username,
                                          password: password,
                                          host: host

def bytes_to_tb(metric) # Pure reports capacity in bytes
  (metric.to_f / 1_099_511_627_776).round(3)
end

count = 0
Dir.glob("#{base_dir}log_capacity/*.csv") do |file|
  # break if count == 20000 #uncomment to restrict number of files processed
  if File.zero?(file) # Ignore empty files and move them to "bad" directory
    puts
    puts 'Empty, moving to bad directory, file: ' + file
    puts
    File.rename(file, "#{base_dir}bad/#{File.basename(file)}.empty")
    next
  end
  timey = DateTime.parse(File.basename(file)[/_(.*)\./, 1]).strftime('%s')
  ray   = File.basename(file)[/^(.*)_/, 1]
  case ray.downcase
  when /^sa0\d{3}ps\d{2}/
    tags = { array: ray, type: 'Pure' }
  when /^sa0\d{3}xms\d{2}/
    tags = { array: ray, type: 'Xtremio' }
  when /^1306/, /^1765/, /^1835/
    tags = { array: ray, type: 'Vmax' }
  end
  values = {}
  CSV.foreach(file) do |row|
    values[row[0]] = row[1]
  end
  x = values['total'].to_f / values['capacity'].to_f
  values['pct_used'] = (x * 100.00).round(3)
  %w(capacity system snapshots volumes total shared_space).each do |metric|
    values[metric] = bytes_to_tb(values[metric])
  end
  %w(data_reduction thin_provisioning total_reduction).each do |metric|
    values[metric] = values[metric].to_f.round(3)
  end
  data = {
    values: values,
    tags: tags,
    timestamp: timey
  }
  begin
    influxdb.write_point(measure, data)
  rescue
    puts
    puts 'Unable to push to database, moving to bad directory, file: ' + file
    puts
    File.rename(file, "#{base_dir}bad/#{File.basename(file)}.dbad")
    next
  end
  sleep(0.5)
  File.rename(file, "#{base_dir}old/#{File.basename(file)}")
  count += 1
  print count.to_s + '.'
end
puts
puts 'Processed ' + count.to_s + ' files.'


storage/influxxer_vmax.rb
#!/usr/bin/env ruby
# Processes csv files containing historical Vmax array capacity data.

require 'csv'
require 'json'
require 'influxdb'

conf     = JSON.parse(File.read('/u01/app/prd/vmax_stats/vmax_stats.json'))
base_dir = conf['base_dir']
database = conf['database']
host     = conf['db_host']
username = conf['db_user']
password = Base64.decode64(conf['db_pass'])
measure  = 'capacity' # The metric

influxdb = InfluxDB::Client.new database, username: username,
                                          password: password,
                                          host: host

count = 0
Dir.glob("#{base_dir}log_capacity/*.csv") do |file|
  # break if count == 20000 #uncomment to restrict number of files processed
  if File.zero?(file) # Ignore empty files and move them to "bad" directory
    puts
    puts 'Empty, moving to bad directory, file: ' + file
    puts
    File.rename(file, "#{base_dir}bad/#{File.basename(file)}.empty")
    next
  end
  timey = DateTime.parse(File.basename(file)[/_(.*)\./, 1]).strftime('%s')
  ray   = File.basename(file)[/^(.*)_/, 1]
  case ray.downcase
  when /^sa0\d{3}ps\d{2}/
    tags = { array: ray, type: 'Pure' }
  when /^sa0\d{3}xms\d{2}/
    tags = { array: ray, type: 'Xtremio' }
  when /^1306/, /^1765/, /^1835/
    tags = { array: ray, type: 'Vmax' }
  end
  values = {}
  info = CSV.read(file)
  values['pct_used']      = (info.last[-3].to_f * 1.00).round(3)
  values['subs_percent']  = info.last.last.to_f
  values['total_used_tb'] = (info.last[-6].to_f * 1.00).round(3)
  values['total_free_tb'] = (info.last[-5].to_f * 1.00).round(3)
  data = {
    values: values,
    tags: tags,
    timestamp: timey
  }
  begin
    influxdb.write_point(measure, data)
  rescue
    puts
    puts 'Unable to push to database, moving to bad directory, file: ' + file
    puts
    File.rename(file, "#{base_dir}bad/#{File.basename(file)}i.dbad")
    next
  end
  sleep(0.5)
  File.rename(file, "#{base_dir}old/#{File.basename(file)}")
  count += 1
  print count.to_s + '.'
end
puts
puts 'Processed ' + count.to_s + ' files.'


storage/perf_tracer_pure.rb
#!/usr/bin/env ruby
# Captures Pure performance metrics and sends to InfluxDB.
# rubocop:disable Metrics/LineLength

require 'csv'
require 'json'
require 'influxdb'
require 'stoarray'

conf     = JSON.parse(File.read(File.join(File.dirname(__FILE__), "pure_stats.json")))
database = conf['database']
host     = conf['db_host']
username = conf['db_user']
password = Base64.decode64(conf['db_pass'])
measure  = 'performance' # The metric

influxdb = InfluxDB::Client.new database, username: username,
                                          password: password,
                                          host: host

def bytes_to_mb(metric)
  metric.to_i / 1_048_576
end

def micro_to_milli(metric)
  (metric.to_f / 1000).round(3)
end

conf['arrays'].each do |ray, key|
  base_url = 'https://' + ray + '.nordstrom.net/api/1.4/'
  p_url     = base_url + 'auth/session'
  p_headers = conf['headers']
  token     = Base64.decode64(key)
  params    = { api_token: token }
  cookies   = Stoarray.new(headers: p_headers, meth: 'Post', params: params, url: p_url).cookie
  p_headers['Cookie'] = cookies
  p_url = base_url + 'array?action=monitor'
  pstat = Stoarray.new(headers: p_headers, meth: 'Get', params: {}, url: p_url).array
  tags = { array: ray, type: 'Pure' }
  values = pstat['response'][0]
  %w(output_per_sec input_per_sec).each do |metric|
    values[metric] = bytes_to_mb(values[metric])
  end
  %w(usec_per_write_op usec_per_read_op).each do |metric|
    values[metric] = micro_to_milli(values[metric])
  end
  %w(writes_per_sec reads_per_sec queue_depth).each do |metric|
    values[metric] = values[metric].to_i
  end
  values.delete('time') # Removing timestamp and changing usec to ms
  values['ms_per_write_op']    = values.delete('usec_per_write_op')
  values['ms_per_read_op']     = values.delete('usec_per_read_op')
  data = {
    values: values,
    tags: tags
  }
  influxdb.write_point(measure, data)
  sleep(0.1)
end


storage/perf_tracer_vmax.rb
#!/usr/bin/env ruby
# Captures Vmax performance metrics and sends to InfluxDB.
# Some code is from Craig Smith's VMAX-Graphite repo, link below:
# https://github.com/nctiggy/VMAX-Graphite/blob/master/collector.rb
# rubocop:disable Metrics/LineLength

require 'crack'
require 'influxdb'
require 'json'
require 'open3'
require 'stoarray'
require 'rest-client'

conf     = JSON.parse(File.read(File.join(File.dirname(__FILE__), "vmax_stats.json")))
database = conf['database']
host     = conf['db_host']
username = conf['db_user']
password = Base64.decode64(conf['db_pass'])
measure  = 'performance' # The metric

influxdb = InfluxDB::Client.new database, username: username,
                                          password: password,
                                          host: host

# Begin code from Craig Smith's VMAX-Graphite
####################################################################################
# Method: Read's the Unisphere XSD file and gets all Metrics for the specified scope
####################################################################################
def get_metrics(param_type,xsd)
  output = Array.new
  JSON.parse(xsd)['xs:schema']['xs:simpleType'].each do |type|
    if type['name'] == "#{param_type}Metric"
      type['xs:restriction']['xs:enumeration'].each do |metric|
        output.push(metric['value']) if metric['value'] == metric['value'].upcase
      end
    end
  end
  return output
end

#####################################
# Method: Reutrns keys for all scopes
#####################################
def get_keys(unisphere,payload,monitor,auth)
  if monitor['scope'].downcase == "array"
    rest = rest_get("https://#{unisphere['ip']}:#{unisphere['port']}/univmax/restapi/performance/#{monitor['scope']}/keys", auth)
  else
    rest = rest_post(payload.to_json,"https://#{unisphere['ip']}:#{unisphere['port']}/univmax/restapi/performance/#{monitor['scope']}/keys", auth)
  end
  componentId = get_component_id_payload(monitor['scope'])
  output = rest["#{componentId}Info"]
  return output
end

##################################################
# Method: Find differences in the key payload
##################################################
def diff_key_payload(incoming_payload,parent_id=nil)
  baseline_keys=["firstAvailableDate","lastAvailableDate"]
  incoming_keys=incoming_payload.keys
  return incoming_keys-baseline_keys
end

##################################################
# Method: Build the Key Payload
##################################################
def build_key_payload(unisphere,symmetrix,monitor,key=nil,parent_id=nil)
  payload = { "symmetrixId" => symmetrix['sid']}
  extra_payload = {parent_id[0] => key[parent_id[0]]} if parent_id
  payload.merge!(extra_payload) if parent_id
  return payload
end

##################################################
# Method: Build the Metric Payload
##################################################
def build_metric_payload(unisphere,monitor,symmetrix,metrics,key=nil,parent_id=nil,child_key=nil,child_id=nil)
  payload = { "symmetrixId" => symmetrix['sid'], "metrics" => metrics}
  parent_payload = { parent_id[0] => key[parent_id[0]] } unless monitor['scope'] == "Array"
  payload.merge!(parent_payload) unless monitor['scope'] == "Array"
  child_payload = { child_id[0] => child_key[child_id[0]], "startDate" => child_key['lastAvailableDate'], "endDate" => child_key['lastAvailableDate'] } if child_key
  payload.merge!(child_payload) if child_key
  timestamp_payload = { "startDate" => key['lastAvailableDate'], "endDate" => key['lastAvailableDate'] } unless child_key
  payload.merge!(timestamp_payload) unless child_key
  uni8_payload = { "dataFormat" => "Average" }
  payload.merge!(uni8_payload)
  return payload
end

################################################################################
# Method: Returns Metrics for all component scopes. Helper for building payloads
################################################################################
def get_perf_metrics(unisphere,payload,monitor,auth)
  rest = rest_post(payload.to_json,"https://#{unisphere['ip']}:#{unisphere['port']}/univmax/restapi/performance/#{monitor['scope']}/metrics", auth)
  output = rest['resultList']['result'][0]
  return output
end

#########################
# Method: API Post Method
#########################
def rest_post(payload, api_url, auth, cert=nil)
  JSON.parse(RestClient::Request.execute(
    method: :post,
    url: api_url,
    verify_ssl: false,
    payload: payload,
    headers: {
      authorization: auth,
      content_type: 'application/json',
      accept: :json
    }
  ))
end

##################################################################################
# Method: To correctly format scope for JSON
##################################################################################
def get_component_id_payload(scope)
  s = scope.split /(?=[A-Z])/
  i = 0
  if s[-1].capitalize == "Pool"
    new_scope = "pool"
  else
    while i < s.length
      s[i] = s[i].downcase if s[i] == s[i].upcase
      s[i] = s[i].downcase if i == 0 && s[i] == s[i].capitalize
      i += 1
    end
    new_scope = s.join
  end
  return new_scope
end

########################
# Method: API GET Method
########################
def rest_get(api_url, auth, cert=nil)
  JSON.parse(RestClient::Request.execute(method: :get,
    url: api_url,
    verify_ssl: false,
    headers: {
      authorization: auth,
      accept: :json
    }
  ))
end

conf['unisphere'].each do |unisphere|
  ## Read the appropriate XSD file based on unisphere version ##
  myparams = Crack::XML.parse(File.read(File.join(File.dirname(__FILE__), "unisphere#{unisphere['version']}.xsd"))).to_json
  ## Build the Base64 auth string ##
  auth = Base64.strict_encode64("#{unisphere['user']}:#{unisphere['password']}")
  ## Loop through each symmetrix in the current unisphere ##
  unisphere['symmetrix'].each do |symmetrix|
    output_payload = {}
    ## Loop through each component on the current symmetrix ##
    conf['monitor'].each do |monitor|
      ## If the component is not enabled i.e. false then skip. If the parent is false it will skip the children ##
      next unless monitor['enabled']
      metrics_param = get_metrics(monitor['scope'],myparams)
      key_payload = build_key_payload(unisphere,symmetrix,monitor)
      keys = get_keys(unisphere,key_payload,monitor,auth)
      keys.each do |key|
        parent_ids = diff_key_payload(key)
        if monitor.key?("children")
          if monitor['children'][0]['enabled']
            child_payload = build_key_payload(unisphere,symmetrix,monitor['children'][0],key,parent_ids)
            child_keys = get_keys(unisphere,child_payload,monitor['children'][0],auth)
            child_keys.each do |child_key|
              child_ids = diff_key_payload(child_key)
              metrics_param = get_metrics(monitor['children'][0]['scope'],myparams)
              metric_payload = build_metric_payload(unisphere,monitor,symmetrix,metrics_param,key,parent_ids,child_key,child_ids)
              metrics = get_perf_metrics(unisphere,metric_payload,monitor['children'][0],auth)
              metrics_param.each do |metric|
                output_payload[(conf['graphite']['prefix'] ? "#{conf['graphite']['prefix']}." : "") + "symmetrix.#{symmetrix['sid']}.#{monitor['scope']}.#{key[parent_ids[0]]}.#{child_key[child_ids[0]]}.#{metric}"] = metrics[metric]
              end
            end
          end
        end
        if (monitor['scope'] != "Array") || (monitor['scope'] == "Array" && key['symmetrixId'] == symmetrix['sid'])
          metrics_param = get_metrics(monitor['scope'],myparams)
          metric_payload = build_metric_payload(unisphere,monitor,symmetrix,metrics_param,key,parent_ids)
          metrics = get_perf_metrics(unisphere,metric_payload,monitor,auth)
          metrics_param.each do |metric|
            output_payload[metric] = metrics[metric] if monitor['scope'] == "Array"
            output_payload[(conf['graphite']['prefix'] ? "#{conf['graphite']['prefix']}." : "") + "symmetrix.#{symmetrix['sid']}.#{monitor['scope']}.#{key[parent_ids[0]]}.#{metric}"] = metrics[metric] unless monitor['scope'] == "Array"
          end
        end
      end
    end
    # End of code from Craig Smith's VMAX-Graphite. Thank you sir!
    values = {}
    tags = { array: symmetrix['sid'][-4..-1], type: 'Vmax' }
    values['writes_per_sec']  = output_payload['WRITES'].to_f.round(3)
    values['reads_per_sec']   = output_payload['READS'].to_f.round(3)
    values['ms_per_read_op']  = output_payload['RESPONSE_TIME_READ'].to_f.round(3)
    values['ms_per_write_op'] = output_payload['RESPONSE_TIME_WRITE'].to_f.round(3)
    values['input_per_sec']   = output_payload['MB_WRITE_PER_SEC'].to_f.round(3)
    values['output_per_sec']  = output_payload['MB_READ_PER_SEC'].to_f.round(3)
    values['pct_cache_used']  = output_payload['TOTAL_CACHE_UTILIZATION'].to_f.round(3)
    data = {
      values: values,
      tags: tags
    }
    influxdb.write_point(measure, data)
    sleep(0.1)
  end
end


storage/space_tracer_vmax.rb
#!/usr/bin/env ruby
# Captures Vmax capacity metrics and sends to InfluxDB.
# rubocop:disable Metrics/LineLength

require 'base64'
require 'csv'
require 'json'
require 'influxdb'
require 'nokogiri'

conf     = JSON.parse(File.read('/u01/app/prd/vmax_stats/vmax_stats.json'))
base_dir = conf['base_dir']
database = conf['database']
host     = conf['db_host']
username = conf['db_user']
password = Base64.decode64(conf['db_pass'])
measure  = 'capacity' # The metric

influxdb = InfluxDB::Client.new database, username: username,
                                          password: password,
                                          host: host

%w(1306 1765 1835).each do |ray|
  xmsmell = Nokogiri::XML(`/opt/emc/SYMCLI/bin/symcfg -sid #{ray} list -thin -gb -detail -gb -pool -output xml`)
  tags = { array: ray, type: 'Vmax' }
  values = {}
  values['pct_used'] = xmsmell.search('Totals').search('percent_full').text.to_f.round(3)
  values['subs_percent'] = xmsmell.search('Totals').search('subs_percent').text.to_f.round(3)
  values['total_used_tb'] = xmsmell.search('Totals').search('total_used_tracks_tb').text.to_f.round(3)
  values['total_free_tb'] = xmsmell.search('Totals').search('total_free_tracks_tb').text.to_f.round(3)
  values['total_capacity_tb'] = xmsmell.search('Totals').search('total_usable_tracks_tb').text.to_f.round(3)
  data = {
    values: values,
    tags: tags
  }
  influxdb.write_point(measure, data)
  sleep(0.1)
end


storage/perf_tracer_xtremio.rb
#!/usr/bin/env ruby
# Captures Xtremio performance metrics and sends to InfluxDB.
# rubocop:disable Metrics/LineLength

require 'csv'
require 'json'
require 'influxdb'
require 'stoarray'

conf     = JSON.parse(File.read(File.join(File.dirname(__FILE__), "xtremio_stats.json")))
database = conf['database']
host     = conf['db_host']
username = conf['db_user']
password = Base64.decode64(conf['db_pass'])
measure  = 'performance' # The metric

influxdb = InfluxDB::Client.new database, username: username,
                                          password: password,
                                          host: host

def kibibytes_to_mb(metric)
  metric.to_i / 1_024
end

def micro_to_milli(metric)
  (metric.to_f / 1000).round(3)
end

conf['arrays'].each do |key, val|
  base_url = 'https://' + key + '.nordstrom.net/api/json/v2/types/'
  headers = conf['headers']
  headers['authorization'] = val
  ray = key.sub('xms', 'xtm')
  url = base_url + 'clusters?name=' + ray
  perf = Stoarray.new(headers: headers, meth: 'Get', params: {}, url: url).array
  tags = { array: ray, type: 'Xtremio' }
  values = {}
  values['writes_per_sec']  = perf['response']['content']['wr-iops'].to_f.round(3)
  values['reads_per_sec']   = perf['response']['content']['rd-iops'].to_f.round(3)
  values['ms_per_read_op']  = micro_to_milli(perf['response']['content']['rd-latency'])
  values['ms_per_write_op'] = micro_to_milli(perf['response']['content']['wr-latency'])
  values['input_per_sec']   = kibibytes_to_mb(perf['response']['content']['wr-bw'])
  values['output_per_sec']  = kibibytes_to_mb(perf['response']['content']['rd-bw'])
  data = {
    values: values,
    tags: tags
  }
  influxdb.write_point(measure, data)
  sleep(0.1)
end


storage/space_tracer_xtremio.rb
#!/usr/bin/env ruby
# Captures Xtremio capacity metrics and sends to InfluxDB.
# rubocop:disable Metrics/LineLength

require 'csv'
require 'json'
require 'influxdb'
require 'stoarray'

conf     = JSON.parse(File.read(File.join(File.dirname(__FILE__), "xtremio_stats.json")))
database = conf['database']
host     = conf['db_host']
username = conf['db_user']
password = Base64.decode64(conf['db_pass'])
measure  = 'capacity' # The metric

influxdb = InfluxDB::Client.new database, username: username,
                                          password: password,
                                          host: host

def kibi_to_tibi(metric)
  (metric.to_f / 1_073_741_824).round(3)
end

conf['arrays'].each do |key, val|
  base_url = 'https://' + key + '.nordstrom.net/api/json/v2/types/'
  headers = conf['headers']
  headers['authorization'] = val
  ray = key.sub('xms', 'xtm')
  url = base_url + 'clusters?name=' + ray
  capacity = Stoarray.new(headers: headers, meth: 'Get', params: {}, url: url).array
  tags = { array: ray, type: 'Xtremio' }
  values = {}
  values['data_reduction']    = (capacity['response']['content']['logical-space-in-use'].to_f / capacity['response']['content']['ud-ssd-space-in-use'].to_f).round(3)
  values['pct_used']          = ((capacity['response']['content']['ud-ssd-space-in-use'].to_f / capacity['response']['content']['ud-ssd-space'].to_f) * 100.00).round(3)
  values['total_free_tb']     = kibi_to_tibi(capacity['response']['content']['ud-ssd-space'].to_f - capacity['response']['content']['ud-ssd-space-in-use'].to_f)
  values['total_used_tb']     = kibi_to_tibi(capacity['response']['content']['ud-ssd-space-in-use'])
  values['total_capacity_tb'] = kibi_to_tibi(capacity['response']['content']['ud-ssd-space'])
  data = {
    values: values,
    tags: tags
  }
  influxdb.write_point(measure, data)
  sleep(0.1)
end


storage/space_tracer_pure.rb
#!/usr/bin/env ruby
# Captures Pure capacity metrics and sends to InfluxDB.
# rubocop:disable Metrics/LineLength

require 'csv'
require 'json'
require 'influxdb'
require 'stoarray'

conf     = JSON.parse(File.read('/u01/app/prd/pure_stats/pure_stats.json'))
base_dir = conf['base_dir']
database = conf['database']
host     = conf['db_host']
username = conf['db_user']
password = Base64.decode64(conf['db_pass'])
measure  = 'capacity' # The metric

influxdb = InfluxDB::Client.new database, username: username,
                                          password: password,
                                          host: host

def bytes_to_tb(metric)
  (metric.to_f / 1_099_511_627_776).round(3)
end

conf['arrays'].each do |ray, key|
  base_url = 'https://' + ray + '.nordstrom.net/api/1.4/'
  p_url     = base_url + 'auth/session'
  p_headers = conf['headers']
  token     = Base64.decode64(key)
  params    = { api_token: token }
  cookies   = Stoarray.new(headers: p_headers, meth: 'Post', params: params, url: p_url).cookie
  p_headers['Cookie'] = cookies
  p_url = base_url + 'array?space=true'
  astat = Stoarray.new(headers: p_headers, meth: 'Get', params: {}, url: p_url).array
  # Uncomment the line below if you want to also generate a csv file.
  #CSV.open("#{base_dir}log_capacity/#{ray}_#{Time.now.strftime('%Y%m%d%H%M%S')}.csv", 'wb') { |csv| astat['response'].to_a.each { |elem| csv << elem } }
  tags = { array: ray, type: 'Pure' }
  values = {}
  astat['response'].to_a.each do |row|
    values[row[0]] = row[1]
  end
  values['pct_used'] = ((values['total'].to_f / values['capacity'].to_f) * 100.00).round(3)
  %w(capacity system snapshots volumes total shared_space ).each do |metric|
    values[metric] = bytes_to_tb(values[metric])
  end
  %w(data_reduction thin_provisioning total_reduction).each do |metric|
    values[metric] = values[metric].to_f.round(3)
  end
  values['total_used_tb']     = values['total']
  values['total_free_tb']     = (values['capacity'] - values['total']).round(3)
  values['total_capacity_tb'] = values['capacity']
  data = {
    values: values,
    tags: tags
  }
  influxdb.write_point(measure, data)
  sleep(0.1)
end
